ddl/partition.go

```diff
@@ -1206,3 +1208,31 @@ func truncateTableByReassignPartitionIDs(t *meta.Meta, tblInfo *model.TableInfo)
 	tblInfo.Partition.Definitions = newDefs
 	return nil
 }
+
+func onAlterTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
+	var rules []*placement.Rule
+	err := job.DecodeArgs(&rules)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return 0, errors.Trace(err)
+	}
+
+	ver, err := t.GetSchemaVersion()
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	ok, err := infosync.UpdatePlacementRules(nil, rules)
+	if err != nil {
+		if ok {
+			// ok means there is one HTTP resp, but not 200
+			// cancell the job if PD declined us for some reason, avoid lots of meaningless retries
+			job.State = model.JobStateCancelled
+		}
+		return 0, errors.Trace(err)
```

Comment:
Use a more specified error message. It will be exposed to users.

Reply:
https://github.com/pingcap/tidb/pull/18611/commits/2c10e2d61017c4f2ee76abf7d04f97ac4e856a83